deeprank/features/atomicfeature_1.py

```diff
@@ -0,0 +1,703 @@
+
+        # define the dictionaries
+        charge_data = {}
+        charge_data_xyz = {}
+
+        # entire residue or not
+        if extend_contact_to_residue:
+            index_contact_atoms= self.extend_contact_to_residue()
+        else:
+            index_contact_atoms = self.contact_atom_ids
+
+        # loop over the protein
+        for i in index_contact_atoms:
+
+            # atinfo
+            key = tuple(atinfo[i])
+
+            # store in the dicts
+            charge_data[key] = [charge[i]]
+
+            # xyz format
+            chain_dict = [{self.mutant.chain_id: 0}[key[0]]]
+            key = tuple(chain_dict + xyz[i, :].tolist())
+            charge_data_xyz[key] = [charge[i]]
+
+        # add the electrostatic feature
+        self.feature_data['charge'] = charge_data
+        self.feature_data_xyz['charge'] = charge_data_xyz
+
+    ####################################################################
+    #
+    #   PAIR INTERACTIONS
+    #
+    ####################################################################
+
+    def evaluate_pair_interaction(self, print_interactions=False,
+                                  save_interactions=False):
+        """Evalaute the pair interactions (coulomb and vdw).
+
+        Args:
+            print_interactions (bool, optional): print data to screen
+            save_interactions (bool, optional): save the interactions to file.
+        """
+
+        if self.verbose:
+            print('-- Compute interaction energy for contact pairs only')
+
+        # extract information from the pdb2sq
+        xyz = np.array(self.sqldb.get('x,y,z'))
+        atinfo = self.sqldb.get(self.atom_key)
+
+        charge = np.array(self.sqldb.get('CHARGE'))
+        vdw = np.array(self.sqldb.get('eps,sig'))
+        eps, sig = vdw[:, 0], vdw[:, 1]
+
+        # define the dictionaries
+        # these holds data like chainID resname resSeq,name values
+        electro_data = {}
+        vdw_data = {}
+
+        # define the dict that hold
+        #  x y z values
+        electro_data_xyz = {}
+        vdw_data_xyz = {}
+
+        # define the matrices
+        nat = len(self.sqldb.get('x', chainID=self.mutant.chain_id))
+        matrix_elec = np.zeros(nat)
+        matrix_vdw = np.zeros(nat)
+
+        # handle the export of the interaction breakdown
+        _save_ = False
+        if save_interactions:
+            if save_interactions:
+                save_interactions = './'
+            if os.path.isdir(save_interactions):
+                fname = os.path.join(save_interactions,
+                                     'atomic_pair_interaction.dat')
+            else:
+                fname = save_interactions
+            f = open(fname, 'w')
+            _save_ = True
+
+        # total energy terms
+        ec_tot, evdw_tot = 0, 0
+
+        # loop over the given chain 
+        #for every contact pair A and B, note the order
+        for iA, iB in self.contact_atoms.items():
```

Comment:
I can't find a point in the code where "self.contact_atoms" is set. Doesn't this throw an exception?

Reply:
Corrected